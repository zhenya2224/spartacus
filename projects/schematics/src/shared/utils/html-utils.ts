import { SchematicsException, Tree } from '@angular-devkit/schematics';
import {
  DefaultTreeDocument,
  DefaultTreeElement,
  DefaultTreeNode,
  parse as parseHtml,
} from 'parse5';

/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 *
 * see: https://github.com/angular/components/blob/master/src/cdk/schematics/utils/html-manipulation.ts
 */
/** Appends the given element HTML fragment to the `<head>` element of the specified HTML file. */
export function appendHtmlElementToHead(
  host: Tree,
  htmlFilePath: string,
  elementHtml: string
) {
  const htmlFileBuffer = host.read(htmlFilePath);

  if (!htmlFileBuffer) {
    throw new SchematicsException(
      `Could not read file for path: ${htmlFilePath}`
    );
  }

  const htmlContent = htmlFileBuffer.toString();

  if (htmlContent.includes(elementHtml)) {
    return;
  }

  const headTag = getHtmlHeadTagElement(htmlContent);

  if (!headTag) {
    throw Error(
      `Could not find '<head>' element in HTML file: ${htmlFileBuffer}`
    );
  }

  // We always have access to the source code location here because the `getHeadTagElement`
  // function explicitly has the `sourceCodeLocationInfo` option enabled.
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  const endTagOffset = headTag.sourceCodeLocation!.endTag.startOffset;
  const indentationOffset = getChildElementIndentation(headTag);
  const insertion = `${' '.repeat(indentationOffset)}${elementHtml}`;

  const recordedChange = host
    .beginUpdate(htmlFilePath)
    .insertRight(endTagOffset, `${insertion}\n`);

  host.commitUpdate(recordedChange);
}
/** Parses the given HTML file and returns the head element if available. */
function getHtmlHeadTagElement(htmlContent: string): DefaultTreeElement | null {
  return getElementByTagName('head', htmlContent);
}
/** Finds an element by its tag name. */
function getElementByTagName(
  tagName: string,
  htmlContent: string
): DefaultTreeElement | null {
  const document = parseHtml(htmlContent, {
    sourceCodeLocationInfo: true,
  }) as DefaultTreeDocument;
  const nodeQueue = [...document.childNodes];

  while (nodeQueue.length) {
    const node = nodeQueue.shift() as DefaultTreeElement;

    if (node.nodeName.toLowerCase() === tagName) {
      return node;
    } else if (node.childNodes) {
      nodeQueue.push(...node.childNodes);
    }
  }

  return null;
}

/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 *
 * See https://github.com/angular/components/blob/master/src/cdk/schematics/utils/parse5-element.ts
 */
/** Determines the indentation of child elements for the given Parse5 element. */
function getChildElementIndentation(element: DefaultTreeElement) {
  const childElement = element.childNodes.find((node: DefaultTreeNode) => {
    return !!(node as any).tagName;
  }) as DefaultTreeElement | null;

  if (
    (childElement && !childElement.sourceCodeLocation) ||
    !element.sourceCodeLocation
  ) {
    throw new SchematicsException(
      'Cannot determine child element indentation because the ' +
        'specified Parse5 element does not have any source code location metadata.'
    );
  }

  const startColumns = childElement
    ? // In case there are child elements inside of the element, we assume that their
      // indentation is also applicable for other child elements.
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      childElement.sourceCodeLocation!.startCol
    : // In case there is no child element, we just assume that child elements should be indented
      // by two spaces.
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      element.sourceCodeLocation!.startCol + 2;

  // Since Parse5 does not set the `startCol` properties as zero-based, we need to subtract
  // one column in order to have a proper zero-based offset for the indentation.
  return startColumns - 1;
}
