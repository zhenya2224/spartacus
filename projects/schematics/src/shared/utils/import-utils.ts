import { Identifier, ts } from 'ts-morph';
import { SPARTACUS_SCOPE } from '../constants';

export function isImportedFrom(node: Identifier, importPath: string): boolean {
  const definitions = node.getDefinitions();
  for (const def of definitions) {
    const node = def.getDeclarationNode();

    const declaration = node?.getFirstAncestorByKind(
      ts.SyntaxKind.ImportDeclaration
    );
    if (declaration?.getModuleSpecifier().getText().includes(importPath)) {
      return true;
    }
  }

  return false;
}

export function isImportedFromSpartacusLibs(node: Identifier): boolean {
  const definitions = node.getDefinitions();
  for (const def of definitions) {
    const node = def.getDeclarationNode();

    const declaration = node?.getFirstAncestorByKind(
      ts.SyntaxKind.ImportDeclaration
    );
    const moduleSpecifier = declaration?.getModuleSpecifier().getText();
    if (
      moduleSpecifier?.startsWith(`'${SPARTACUS_SCOPE}`) ||
      moduleSpecifier?.startsWith(`"${SPARTACUS_SCOPE}`)
    ) {
      return true;
    }
  }

  return false;
}
