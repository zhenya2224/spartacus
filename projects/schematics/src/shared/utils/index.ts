export * from './config-utils';
export * from './file-utils';
export * from './lib-utils';
export * from './module-file-utils';
export * from './new-module-utils';
export * from './package-utils';
export * from './program';
export * from './project-tsconfig-paths';
export * from './workspace-utils';
