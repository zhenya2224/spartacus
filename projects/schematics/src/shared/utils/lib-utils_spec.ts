import {
  SchematicTestRunner,
  UnitTestTree,
} from '@angular-devkit/schematics/testing';
import {
  Schema as ApplicationOptions,
  Style,
} from '@schematics/angular/application/schema';
import { Schema as WorkspaceOptions } from '@schematics/angular/workspace/schema';
import * as path from 'path';
import { Schema as SpartacusOptions } from '../../add-spartacus/schema';
import { UTF_8 } from '../constants';
import {
  addLibraryFeature,
  FeatureConfig,
  LibraryOptions,
  shouldAddFeature,
} from './lib-utils';

const collectionPath = path.join(__dirname, '../../collection.json');

const CLI_FEATURE_NAME = 'xxx-cli';
const FEATURE_NAME = 'xxx';
const FEATURE_FOLDER_NAME = 'xxx';
const FEATURE_MODULE_NAME = 'XxxModule';
const FEATURE_MODULE_IMPORT_PATH = '@spartacus/xxx';
const ROOT_MODULE_NAME = 'XxxModuleRoot';
const ROOT_FEATURE_MODULE_IMPORT_PATH = '@spartacus/xxx/root';
const I18N_RESOURCES = 'translations';
const I18N_CHUNKS = 'translationChunk';
const ASSETS_IMPORT_PATH = '@spartacus/xxx/assets';
const SCSS_FILE_NAME = 'xxx.scss';
const STYLE_IMPORT_PATH = FEATURE_MODULE_IMPORT_PATH;

const appModulePath = 'src/app/app.module.ts';
const scssFilePath = `src/styles/spartacus/${SCSS_FILE_NAME}`;

const BASE_FEATURE_CONFIG: FeatureConfig = {
  moduleName: FEATURE_NAME,
  folderName: FEATURE_FOLDER_NAME,
  featureModule: {
    name: FEATURE_MODULE_NAME,
    importPath: FEATURE_MODULE_IMPORT_PATH,
  },
  rootModule: {
    name: ROOT_MODULE_NAME,
    importPath: ROOT_FEATURE_MODULE_IMPORT_PATH,
  },
  i18n: {
    resources: I18N_RESOURCES,
    chunks: I18N_CHUNKS,
    importPath: ASSETS_IMPORT_PATH,
  },
  styles: {
    scssFileName: SCSS_FILE_NAME,
    importStyle: STYLE_IMPORT_PATH,
  },
};

const BASE_OPTIONS: LibraryOptions = {
  project: 'schematics-test',
  features: [CLI_FEATURE_NAME],
  lazy: true,
};

describe('Lib utils', () => {
  describe('shouldAddFeature', () => {
    it('should return true if the feature is present in the given features array', () => {
      const feature1 = 'feature1';
      const features = [feature1];
      expect(shouldAddFeature(feature1, features)).toBe(true);
    });
    it('should return false if the feature is NOT present in the given features array', () => {
      const random = 'random';
      const feature1 = 'feature1';
      const features = [feature1];
      expect(shouldAddFeature(random, features)).toBe(false);
    });
  });

  describe('addLibraryFeature', () => {
    const schematicRunner = new SchematicTestRunner(
      'schematics',
      collectionPath
    );

    let appTree: UnitTestTree;

    const workspaceOptions: WorkspaceOptions = {
      name: 'workspace',
      version: '0.5.0',
    };

    const appOptions: ApplicationOptions = {
      name: 'schematics-test',
      inlineStyle: false,
      inlineTemplate: false,
      routing: false,
      style: Style.Scss,
      skipTests: false,
      projectRoot: '',
    };

    const spartacusDefaultOptions: SpartacusOptions = {
      project: 'schematics-test',
      lazy: true,
      features: [],
    };

    beforeEach(async () => {
      appTree = await schematicRunner
        .runExternalSchematicAsync(
          '@schematics/angular',
          'workspace',
          workspaceOptions
        )
        .toPromise();
      appTree = await schematicRunner
        .runExternalSchematicAsync(
          '@schematics/angular',
          'application',
          appOptions,
          appTree
        )
        .toPromise();
      appTree = await schematicRunner
        .runSchematicAsync(
          'add-spartacus',
          { ...spartacusDefaultOptions, name: 'schematics-test' },
          appTree
        )
        .toPromise();
    });

    it('should add i18n config in feature module', async () => {
      const rule = addLibraryFeature(BASE_OPTIONS, BASE_FEATURE_CONFIG);
      const tree = await schematicRunner.callRule(rule, appTree).toPromise();

      // TODO: Finish when config util will be created
      const appModule = tree.read(appModulePath)?.toString(UTF_8);
      expect(appModule).not.toContain(
        `import { ${I18N_RESOURCES} } from '${ASSETS_IMPORT_PATH}';`
      );
    });
    it('should NOT add i18n if the config is not present', async () => {
      const featureConfig: FeatureConfig = {
        ...BASE_FEATURE_CONFIG,
        i18n: undefined,
      };
      const rule = addLibraryFeature(BASE_OPTIONS, featureConfig);
      const tree = await schematicRunner.callRule(rule, appTree).toPromise();

      // TODO: Finish when config util will be created
      const appModule = tree.read(appModulePath)?.toString(UTF_8);
      expect(appModule).not.toContain(`providers: [
        provideConfig({
          i18n: {`);
    });
    describe('when the lazy loading is configured', () => {
      it('should add it in the lazy loading way', async () => {
        const rule = addLibraryFeature(BASE_OPTIONS, BASE_FEATURE_CONFIG);
        const tree = await schematicRunner.callRule(rule, appTree).toPromise();

        const appModule = tree.read(appModulePath)?.toString(UTF_8);
        // TODO: Finish when config util will be created
        expect(appModule).not.toContain(
          `import { ${ROOT_MODULE_NAME} } from '${ROOT_FEATURE_MODULE_IMPORT_PATH}';`
        );
      });
    });
    describe('when the eager loading is configured', () => {
      it('should add it in the eager way', async () => {
        const rule = addLibraryFeature(
          { ...BASE_OPTIONS, lazy: false },
          BASE_FEATURE_CONFIG
        );
        const tree = await schematicRunner.callRule(rule, appTree).toPromise();

        const appModule = tree.read(appModulePath)?.toString(UTF_8);
        // TODO: Finish when config util will be created
        expect(appModule).not.toContain(
          `module: () => import('${FEATURE_MODULE_IMPORT_PATH}').then(`
        );
      });
    });
    describe('style', () => {
      describe('when style config is provided', () => {
        describe('and the scss file does NOT exist', () => {
          it('should add it', async () => {
            const rule = addLibraryFeature(BASE_OPTIONS, BASE_FEATURE_CONFIG);
            const tree = await schematicRunner
              .callRule(rule, appTree)
              .toPromise();

            expect(tree.exists(scssFilePath)).toEqual(true);
            const content = tree.read(scssFilePath)?.toString(UTF_8);
            expect(content).toEqual(`@import "${FEATURE_MODULE_IMPORT_PATH}";`);
          });
        });
        describe('and the scss with the same content already exists', () => {
          beforeEach(() => {
            appTree.create(
              scssFilePath,
              `@import "${FEATURE_MODULE_IMPORT_PATH}";`
            );
          });
          it('should NOT append it', async () => {
            const rule = addLibraryFeature(BASE_OPTIONS, BASE_FEATURE_CONFIG);
            const tree = await schematicRunner
              .callRule(rule, appTree)
              .toPromise();

            expect(tree.exists(scssFilePath)).toEqual(true);
            const content = tree.read(scssFilePath)?.toString(UTF_8);
            expect(content).toEqual(`@import "${FEATURE_MODULE_IMPORT_PATH}";`);
          });
        });
        describe('and the scss file with a different content already exists', () => {
          const randomContent = `@import "@random/xxx";`;
          beforeEach(() => {
            appTree.create(scssFilePath, randomContent);
          });
          it('should append it', async () => {
            const rule = addLibraryFeature(BASE_OPTIONS, BASE_FEATURE_CONFIG);
            const tree = await schematicRunner
              .callRule(rule, appTree)
              .toPromise();

            expect(tree.exists(scssFilePath)).toEqual(true);
            const content = tree.read(scssFilePath)?.toString(UTF_8);
            expect(content).toEqual(
              `${randomContent}\n@import "${FEATURE_MODULE_IMPORT_PATH}";`
            );
          });
        });
      });
      describe('when style config is NOT provided', () => {
        it('should not add it', async () => {
          const rule = addLibraryFeature(BASE_OPTIONS, {
            ...BASE_FEATURE_CONFIG,
            styles: undefined,
          });
          const tree = await schematicRunner
            .callRule(rule, appTree)
            .toPromise();

          expect(tree.exists(scssFilePath)).toEqual(false);
        });
      });
    });
  });
});
