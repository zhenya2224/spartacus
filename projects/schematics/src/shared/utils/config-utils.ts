import { Node, SourceFile, ts as tsMorph } from 'ts-morph';
import {
  PROVIDE_CONFIG_FUNCTION,
  SPARTACUS_CORE,
  SPARTACUS_SETUP,
} from '../constants';
import { isImportedFromSpartacusLibs } from './import-utils';
import { CustomConfig } from './lib-utils';
import { getModule } from './new-module-utils';

export function getSpartacusProviders(sourceFile: SourceFile): Node[] {
  const module = getModule(sourceFile);
  if (!module) {
    return [];
  }
  const literal = module.getFirstDescendantByKind(
    tsMorph.SyntaxKind.ObjectLiteralExpression
  );
  const providers: Node[] = [];

  const properties =
    literal?.getChildrenOfKind(tsMorph.SyntaxKind.PropertyAssignment) ?? [];
  properties.forEach((property) => {
    if (
      property.getNameNode().getText() === 'providers' &&
      property.getInitializerIfKind(tsMorph.SyntaxKind.ArrayLiteralExpression)
    ) {
      const initializer = property.getInitializerIfKind(
        tsMorph.SyntaxKind.ArrayLiteralExpression
      );
      initializer?.getElements().forEach((element) => {
        if (Node.isCallExpression(element) || Node.isSpreadElement(element)) {
          const expression = element.getExpression();
          if (
            Node.isIdentifier(expression) &&
            isImportedFromSpartacusLibs(expression)
          ) {
            providers.push(element);
          }
        }
      });
    }
  });

  return providers;
}

const EMPTY_SPACE_REG_EXP = /\s+/gm;
export function normalizeConfiguration(config: string | Node): string {
  let newConfig = typeof config === 'string' ? config : config.getText();

  newConfig = newConfig.trim();

  if (newConfig.startsWith(PROVIDE_CONFIG_FUNCTION)) {
    newConfig = newConfig.replace(`${PROVIDE_CONFIG_FUNCTION}(`, '');
    newConfig = newConfig.substring(0, newConfig.length - 1);
  }

  newConfig = newConfig.replace(EMPTY_SPACE_REG_EXP, '');

  return newConfig;
}

export function getB2bConfiguration(): CustomConfig[] {
  return [
    {
      import: [
        {
          moduleSpecifier: SPARTACUS_CORE,
          namedImports: [PROVIDE_CONFIG_FUNCTION],
        },
        {
          moduleSpecifier: SPARTACUS_SETUP,
          namedImports: ['defaultB2bOccConfig'],
        },
      ],
      content: `provideConfig(defaultB2bOccConfig)`,
    },
    {
      import: [
        {
          moduleSpecifier: SPARTACUS_CORE,
          namedImports: [PROVIDE_CONFIG_FUNCTION],
        },
        {
          moduleSpecifier: SPARTACUS_SETUP,
          namedImports: ['defaultB2bCheckoutConfig'],
        },
      ],
      content: `provideConfig(defaultB2bCheckoutConfig)`,
    },
  ];
}
