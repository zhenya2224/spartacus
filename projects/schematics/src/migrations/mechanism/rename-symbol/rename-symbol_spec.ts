/// <reference types="jest" />

import { getSystemPath } from '@angular-devkit/core';
import { TempScopedNodeJsSyncHost } from '@angular-devkit/core/node/testing';
import { HostTree, Tree } from '@angular-devkit/schematics';
import {
  SchematicTestRunner,
  UnitTestTree,
} from '@angular-devkit/schematics/testing';
import * as shx from 'shelljs';
import { runMigration, writeFile } from '../../../shared/utils/test-utils';

const MIGRATION_SCRIPT_NAME = '00-migration-v4-rename-symbol';

const fileWithSimpleImport = `import { OtherComponent1, Test1Component } from "@spartacus/storefront";
import { Test2Component } from "@spartacus/core";

const array = [OtherComponent1, Test1Component, Test2Component];`;

// -----------------------------------------------------------------------

const fileWithSimpleImportWithAlias = `import { OtherComponent2 as Test, Test1Component } from "@spartacus/storefront";
import { Test2Component } from "@spartacus/core";

const array = [Test, Test1Component, Test2Component];`;

// -----------------------------------------------------------------------

const fileWithSimpleImportAndRename = `import { OtherComponent3, Test1Component } from "@spartacus/storefront";
import { Test2Component } from "@spartacus/core";

const array = [OtherComponent3, Test1Component, Test2Component];`;
// -----------------------------------------------------------------------

const fileWithRename = `import { OtherComponent4, Test1Component } from "@spartacus/storefront";
import { Test2Component } from "@spartacus/core";

const array = [OtherComponent4, Test1Component, Test2Component];`;
// -----------------------------------------------------------------------

const fileWithImportsAndCtor = `
import { Injectable } from '@angular/core';
import {
  AsmAuthHttpHeaderService,
  AuthService,
  AuthStorageService,
  CsAgentAuthService,
  GlobalMessageService,
  OAuthLibWrapperService,
  OccEndpointsService,
  RoutingService,
} from '@spartacus/core';

@Injectable({
  providedIn: 'root',
})
export class X extends AsmAuthHttpHeaderService {
  constructor(
    protected authService: AuthService,
    protected authStorageService: AuthStorageService,
    protected csAgentAuthService: CsAgentAuthService,
    protected oAuthLibWrapperService: OAuthLibWrapperService,
    protected routingService: RoutingService,
    protected globalMessageService: GlobalMessageService,
    protected occEndpointsService: OccEndpointsService
  ) {
    super(
      authService,
      authStorageService,
      csAgentAuthService,
      oAuthLibWrapperService,
      routingService,
      globalMessageService,
      occEndpointsService
    );
  }
}
`;

describe('renamed symbols', () => {
  let host: TempScopedNodeJsSyncHost;
  let appTree = Tree.empty() as UnitTestTree;
  let schematicRunner: SchematicTestRunner;
  let tmpDirPath: string;
  let previousWorkingDir: string;

  beforeEach(() => {
    schematicRunner = new SchematicTestRunner(
      'test',
      require.resolve('../../test/migrations-test.json')
    );
    host = new TempScopedNodeJsSyncHost();
    appTree = new UnitTestTree(new HostTree(host));

    writeFile(
      host,
      '/tsconfig.json',
      JSON.stringify({
        compilerOptions: {
          lib: ['es2015'],
        },
      })
    );

    writeFile(
      host,
      '/angular.json',
      JSON.stringify({
        projects: {
          'spartacus-test': {
            sourceRoot: 'src',
            architect: {
              build: { options: { tsConfig: './tsconfig.json' } },
            },
          },
        },
      })
    );

    previousWorkingDir = shx.pwd();
    tmpDirPath = getSystemPath(host.root);

    // Switch into the temporary directory path. This allows us to run
    // the schematic against our custom unit test tree.
    shx.cd(tmpDirPath);
  });

  afterEach(() => {
    shx.cd(previousWorkingDir);
    shx.rm('-r', tmpDirPath);
  });

  describe('Previous import', () => {
    it('should became new import', async () => {
      writeFile(host, '/src/index.ts', fileWithSimpleImport);

      await runMigration(appTree, schematicRunner, MIGRATION_SCRIPT_NAME);

      const content = appTree.readContent('/src/index.ts');
      expect(content).toMatchSnapshot();
    });

    it('should became new import and name with aliases', async () => {
      writeFile(host, '/src/index.ts', fileWithSimpleImportWithAlias);

      await runMigration(appTree, schematicRunner, MIGRATION_SCRIPT_NAME);

      const content = appTree.readContent('/src/index.ts');
      expect(content).toMatchSnapshot();
    });

    it('should became new import and new name', async () => {
      writeFile(host, '/src/index.ts', fileWithSimpleImportAndRename);

      await runMigration(appTree, schematicRunner, MIGRATION_SCRIPT_NAME);

      const content = appTree.readContent('/src/index.ts');
      expect(content).toMatchSnapshot();
    });
  });

  it('Should only rename node', async () => {
    writeFile(host, '/src/index.ts', fileWithRename);

    await runMigration(appTree, schematicRunner, MIGRATION_SCRIPT_NAME);

    const content = appTree.readContent('/src/index.ts');
    expect(content).toMatchSnapshot();
  });

  describe('when both rename symbol and ctor migration are required', () => {
    it('should be done in the correct order', async () => {
      writeFile(host, '/src/index.ts', fileWithImportsAndCtor);

      appTree = await runMigration(
        appTree,
        schematicRunner,
        MIGRATION_SCRIPT_NAME
      );
      appTree = await runMigration(
        appTree,
        schematicRunner,
        '01-migration-v4-constructor-deprecations'
      );

      const content = appTree.readContent('/src/index.ts');
      expect(content).toMatchSnapshot();
    });
  });
});
