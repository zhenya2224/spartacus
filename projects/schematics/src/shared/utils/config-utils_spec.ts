/// <reference types="jest" />

import {
  SchematicTestRunner,
  UnitTestTree,
} from '@angular-devkit/schematics/testing';
import {
  Schema as ApplicationOptions,
  Style,
} from '@schematics/angular/application/schema';
import { Schema as WorkspaceOptions } from '@schematics/angular/workspace/schema';
import * as path from 'path';
import { InMemoryFileSystemHost, Project, SourceFile } from 'ts-morph';
import { getSpartacusProviders } from './config-utils';

const collectionPath = path.join(__dirname, '../../collection.json');
const schematicRunner = new SchematicTestRunner('schematics', collectionPath);

describe('Storefront config utils', () => {
  let appTree: UnitTestTree;
  const workspaceOptions: WorkspaceOptions = {
    name: 'workspace',
    version: '0.5.0',
  };
  const appOptions: ApplicationOptions = {
    name: 'schematics-test',
    inlineStyle: false,
    inlineTemplate: false,
    routing: false,
    style: Style.Scss,
    skipTests: false,
    projectRoot: '',
  };
  const defaultOptions = {
    project: 'schematics-test',
  };

  beforeEach(async () => {
    appTree = await schematicRunner
      .runExternalSchematicAsync(
        '@schematics/angular',
        'workspace',
        workspaceOptions
      )
      .toPromise();
    appTree = await schematicRunner
      .runExternalSchematicAsync(
        '@schematics/angular',
        'application',
        appOptions,
        appTree
      )
      .toPromise();
    appTree = await schematicRunner
      .runSchematicAsync('add-spartacus', defaultOptions, appTree)
      .toPromise();
  });

  describe('ts-morph config utils', () => {
    const configFileContent = `
import { NgModule } from '@angular/core';
import {
CartAddEntrySuccessEvent,
CartRemoveEntrySuccessEvent,
provideConfig,
} from '@spartacus/core';
import { NavigationEvent, defaultCmsContentProviders } from '@spartacus/storefront';
import { PersonalizationRootModule } from '@spartacus/tracking/personalization/root';
import { AepModule } from '@spartacus/tracking/tms/aep';
import { BaseTmsModule, TmsConfig } from '@spartacus/tracking/tms/core';
import { GtmModule } from '@spartacus/tracking/tms/gtm';
import { someFunction, someSpread } from '@some/package';

@NgModule({
imports: [
BaseTmsModule.forRoot(),
GtmModule,
AepModule,
PersonalizationRootModule,
],
providers: [
someFunction(),
...someSpread,
...defaultCmsContentProviders,
provideConfig(<TmsConfig>{
  tagManager: {
    gtm: {
      events: [NavigationEvent, CartAddEntrySuccessEvent],
    },
    aep: {
      events: [NavigationEvent, CartRemoveEntrySuccessEvent],
    },
  },
}),
provideConfig({
  featureModules: {
    personalization: {
      module: () =>
        import('@spartacus/tracking/personalization').then(
          (m) => m.PersonalizationModule
        ),
    },
  },
}),
],
})
export class TrackingFeatureModule {}
`;
    let sourceFile: SourceFile;

    beforeAll(() => {
      const project = new Project({
        fileSystem: new InMemoryFileSystemHost(),
      });
      sourceFile = project.createSourceFile('test.ts', configFileContent);
    });

    describe('getSpartacusProviders', () => {
      it('should return all providers from spartacus in file', () => {
        const providers = getSpartacusProviders(sourceFile);
        expect(providers.length).toEqual(3);
        expect(providers[0].getText()).toMatchSnapshot();
        expect(providers[1].getText()).toMatchSnapshot();
        expect(providers[2].getText()).toMatchSnapshot();
      });
    });
  });
});
