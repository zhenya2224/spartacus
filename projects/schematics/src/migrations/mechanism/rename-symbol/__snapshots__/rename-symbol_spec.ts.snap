// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renamed symbols Previous import should became new import 1`] = `
"import { Test2Component } from \\"@spartacus/core\\";
import { OtherComponent1 } from \\"@spartacus/storefinder/components\\";
import { Test1Component } from \\"@spartacus/storefront\\";

const array = [OtherComponent1, Test1Component, Test2Component];
"
`;

exports[`renamed symbols Previous import should became new import and name with aliases 1`] = `
"import { Test2Component } from \\"@spartacus/core\\";
import { OtherComponentTest2 as Test } from \\"@spartacus/storefinder/components\\";
import { Test1Component } from \\"@spartacus/storefront\\";

const array = [Test, Test1Component, Test2Component];
"
`;

exports[`renamed symbols Previous import should became new import and new name 1`] = `
"import { Test2Component } from \\"@spartacus/core\\";
import { OtherComponentTest3 } from \\"@spartacus/storefinder/components\\";
import { Test1Component } from \\"@spartacus/storefront\\";

const array = [OtherComponentTest3, Test1Component, Test2Component];
"
`;

exports[`renamed symbols Should only rename node 1`] = `
"import { Test2Component } from \\"@spartacus/core\\";
import { OtherComponentTest4, Test1Component } from \\"@spartacus/storefront\\";

const array = [OtherComponentTest4, Test1Component, Test2Component];
"
`;

exports[`renamed symbols when both rename symbol and ctor migration are required should be done in the correct order 1`] = `
"
import { Injectable } from '@angular/core';
import { AsmAuthHttpHeaderService, CsAgentAuthService } from \\"@spartacus/asm/root\\";
import {
  AuthService,
  AuthStorageService,
  GlobalMessageService,
  OAuthLibWrapperService,
  OccEndpointsService,
  RoutingService, AuthRedirectService
} from '@spartacus/core';

@Injectable({
  providedIn: 'root',
})
export class X extends AsmAuthHttpHeaderService {
  constructor(
    protected authService: AuthService,
    protected authStorageService: AuthStorageService,
    protected csAgentAuthService: CsAgentAuthService,
    protected oAuthLibWrapperService: OAuthLibWrapperService,
    protected routingService: RoutingService,
    protected globalMessageService: GlobalMessageService,
    protected occEndpointsService: OccEndpointsService, authRedirectService: AuthRedirectService
  ) {
    super(
      authService,
      authStorageService,
      csAgentAuthService,
      oAuthLibWrapperService,
      routingService,
      globalMessageService,
      occEndpointsService, authRedirectService
    );
  }
}
"
`;
